.devcontainer/devcontainer.json

```diff
@@ -0,0 +1,57 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the README at:
+// https://github.com/microsoft/vscode-dev-containers/tree/v0.223.0/containers/python-3
+{
+    "name": "Python 3",
+    "build": {
+        "dockerfile": "Dockerfile",
+        "context": "..",
+        "args": {
+            // Update 'VARIANT' to pick a Python version: 3, 3.10, 3.9, 3.8, 3.7, 3.6
+            // Append -bullseye or -buster to pin to an OS version.
+            // Use -bullseye variants on local on arm64/Apple Silicon.
+            "VARIANT": "3.10-bullseye",
+            // Options
+            "NODE_VERSION": "lts/*"
+        }
+    },
+    // Set *default* container specific settings.json values on container create.
+    "settings": {
+        "python.defaultInterpreterPath": "/usr/local/bin/python",
+        "python.linting.enabled": true,
+        "python.linting.pylintEnabled": true,
+        "python.formatting.autopep8Path": "/usr/local/py-utils/bin/autopep8",
+        "python.formatting.blackPath": "/usr/local/py-utils/bin/black",
+        "python.formatting.yapfPath": "/usr/local/py-utils/bin/yapf",
+        "python.linting.banditPath": "/usr/local/py-utils/bin/bandit",
+        "python.linting.flake8Path": "/usr/local/py-utils/bin/flake8",
+        "python.linting.mypyPath": "/usr/local/py-utils/bin/mypy",
+        "python.linting.pycodestylePath": "/usr/local/py-utils/bin/pycodestyle",
+        "python.linting.pydocstylePath": "/usr/local/py-utils/bin/pydocstyle",
+        "python.linting.pylintPath": "/usr/local/py-utils/bin/pylint",
+        "python.testing.pytestEnabled": true,
+        "restructuredtext.confPath": "${workspaceFolder}/docs",
+        "restructuredtext.pythonRecommendation.disabled": true
+    },
+    // Add the IDs of extensions you want installed when the container is created.
+    "extensions": [
+        "ms-python.python",
+        "ms-python.vscode-pylance",
```

Comment:
I can't find any pylance errors, where did you see them? I'm happy to take a look.

Reply:
Thank you! When you open _base.py, _iter.py, _orderedbase.py, or _orderedbidict.py, do you not see the "problems" flagged that I'm seeing?:

<img width="1600" alt="Screen Shot 2022-03-01 at 10 13 49 AM" src="https://user-images.githubusercontent.com/64992/156195092-1df77169-52d1-4c98-b356-9a76824f253b.png">

Reply:
Weird, no, I don't see any of that, I opened all the same files. I tried this with the normal pylance build and the insiders build. Here I am at the exact spot of one of your reported errors, and I get nothing.

![image](https://user-images.githubusercontent.com/46406/156286206-9b4b542c-aead-40e4-a747-12dd62ff085b.png)

Reply:
Can you go to the output pane and pick the Python Language Server tab and see what logs you get after you run Python: Restart Language Server?

I get this:

```
[Info  - 2:58:44 AM] (2897) Pylance language server 2022.2.4 (pyright 101e717a) starting
[Info  - 2:58:44 AM] (2897) Server root directory: /home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2022.2.4/dist
[Info  - 2:58:44 AM] (2897) No configuration file found.
[Info  - 2:58:44 AM] (2897) pyproject.toml file found at /workspaces/bidict.
[Info  - 2:58:44 AM] (2897) Setting pythonPath for service "bidict": "/usr/local/bin/python"
[Info  - 2:58:44 AM] (2897) Loading pyproject.toml file at /workspaces/bidict/pyproject.toml
[Error - 2:58:44 AM] (2897) Pyproject file "/workspaces/bidict/pyproject.toml" is missing "[tool.pyright]" section.
[Warn  - 2:58:44 AM] (2897) stubPath /workspaces/bidict/typings is not a valid directory.
[Info  - 2:58:44 AM] (2897) Assuming Python version 3.10
[Info  - 2:58:44 AM] (2897) Assuming Python platform Linux
[Warn  - 2:58:45 AM] (2897) Exception received when installing file system watcher: TypeError [ERR_FEATURE_UNAVAILABLE_ON_PLATFORM]: The feature watch recursively is unavailable on the current platform, which is being used to run Node.js
[Warn  - 2:58:45 AM] (2897) Exception received when installing file system watcher: TypeError [ERR_FEATURE_UNAVAILABLE_ON_PLATFORM]: The feature watch recursively is unavailable on the current platform, which is being used to run Node.js
[Warn  - 2:58:45 AM] (2897) Exception received when installing file system watcher: TypeError [ERR_FEATURE_UNAVAILABLE_ON_PLATFORM]: The feature watch recursively is unavailable on the current platform, which is being used to run Node.js
[Warn  - 2:58:45 AM] (2897) Exception received when installing file system watcher: TypeError [ERR_FEATURE_UNAVAILABLE_ON_PLATFORM]: The feature watch recursively is unavailable on the current platform, which is being used to run Node.js
[Info  - 2:58:45 AM] (2897) Searching for source files
[Info  - 2:58:45 AM] (2897) Found 26 source files
[Info  - 2:58:45 AM] (2897) Background analysis(1) root directory: /home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2022.2.4/dist
[Info  - 2:58:45 AM] (2897) Background analysis(1) started
[Info  - 2:58:45 AM] (2897) Indexer background runner(2) root directory: /home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2022.2.4/dist (index)
[Info  - 2:58:45 AM] (2897) Indexing(2) started
[Info  - 2:58:46 AM] (2897) Indexer done(2). indexed 460 files
```
![image](https://user-images.githubusercontent.com/46406/156286496-f8a41b8b-f9c3-45a9-88fd-9e3fd894c78c.png)

Reply:
Ok actually my pylance doesn't seem to be detecting errors it should, my setup is probably weird.

Let's solve this in post?

Reply:
As an experiment, I tried changing VARIANT from "3.10-bullseye" to "3" (in both Dockerfile and devcontainer.json), rebuilding, and it worked. Changed it back to "3.10-bullseye", rebuilt, and... it worked. Ok, so the fact that it's working for me now probably has nothing to do with my faffing with VARIANT, but rather the public key for https://cli.github.com/packages that was missing must no longer be missing? OMG

Reply:
Now that I can get them again, here are my Python Language Server logs, in case they might still be helpful:
```console
[Info  - 6:36:53 PM] (608) Pylance language server 2022.2.4 (pyright 101e717a) starting
[Info  - 6:36:53 PM] (608) Server root directory: /home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2022.2.4/dist
[Info  - 6:36:53 PM] (608) No configuration file found.
[Info  - 6:36:53 PM] (608) pyproject.toml file found at /workspaces/bidict.
[Info  - 6:36:53 PM] (608) Setting pythonPath for service "bidict": "/usr/local/bin/python"
[Info  - 6:36:53 PM] (608) Loading pyproject.toml file at /workspaces/bidict/pyproject.toml
[Error - 6:36:53 PM] (608) Pyproject file "/workspaces/bidict/pyproject.toml" is missing "[tool.pyright]" section.
[Warn  - 6:36:53 PM] (608) stubPath /workspaces/bidict/typings is not a valid directory.
[Info  - 6:36:53 PM] (608) Assuming Python version 3.10
[Info  - 6:36:53 PM] (608) Assuming Python platform Linux
[Warn  - 6:36:53 PM] (608) Exception received when installing file system watcher: TypeError [ERR_FEATURE_UNAVAILABLE_ON_PLATFORM]: The feature watch recursively is unavailable on the current platform, which is being used to run Node.js
[Warn  - 6:36:53 PM] (608) Exception received when installing file system watcher: TypeError [ERR_FEATURE_UNAVAILABLE_ON_PLATFORM]: The feature watch recursively is unavailable on the current platform, which is being used to run Node.js
[Warn  - 6:36:53 PM] (608) Exception received when installing file system watcher: TypeError [ERR_FEATURE_UNAVAILABLE_ON_PLATFORM]: The feature watch recursively is unavailable on the current platform, which is being used to run Node.js
[Warn  - 6:36:53 PM] (608) Exception received when installing file system watcher: TypeError [ERR_FEATURE_UNAVAILABLE_ON_PLATFORM]: The feature watch recursively is unavailable on the current platform, which is being used to run Node.js
[Info  - 6:36:53 PM] (608) Searching for source files
[Info  - 6:36:53 PM] (608) Found 26 source files
[Info  - 6:36:53 PM] (608) Background analysis(1) root directory: /home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2022.2.4/dist
[Info  - 6:36:53 PM] (608) Background analysis(1) started
[Info  - 6:36:54 PM] (608) Indexer background runner(2) root directory: /home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2022.2.4/dist (index)
[Info  - 6:36:54 PM] (608) Indexing(2) started
[Info  - 6:36:55 PM] (608) Indexer done(2). indexed 460 files
```

Reply:
Btw, I just took a quick look for a pylance pre-commit hook and didn’t find one, so I guess it’s not yet possible to ensure that all bundled linters stay in agreement that `main` is lint-free if we keep this VS Code config using pylance by default. 

1. Do you think we’ll see a pylance pre-commit hook in the future? (Is it even designed to be runnable in a headless and one-shot fashion?)
2. What are your thoughts on configuring VS Code here to use something other than pylance, given the above?

Reply:
Hmm, no, that looks like the same as mine, even down to the version/commit.

Reply:
Pylance is just spicy pyright (I believe just extension-specific stuff, the engine and linters are all coming from the open source pyright) and it has a precommit hook: https://github.com/necaris/pre-commit-pyright

Reply:
er, hmm, that's really old, but I bet we can just bump the version number

Reply:
Actually this link is better https://github.com/microsoft/pyright/blob/main/docs/ci-integration.md

Reply:
As we just discussed, we're going with setting VS Code to use Jedi + mypy rather than Pylance for now, to achieve parity with the existing pre-commit setup, and since Pylance's complaints looked bogus. Good to know that if we switch to Pylance in the future, we can add that pyright pre-commit hook. Thanks Leif!

Reply:
I figured out why I didn't get your errors: I needed to set this to `basic` instead of `off`, which is the default: 
![image](https://user-images.githubusercontent.com/46406/156484546-91001a6a-cd28-4216-a266-bc21cb1c409b.png)

I'm not sure why you had it set to `basic`, any chance you did that yourself?

Anyway, if we look at https://github.com/leifwalsh/bidict/blob/5d8f520144e70b4da5fdcdd5f9e223f3e8e13e71/bidict/_base.py#L134, pyright is basically saying "I can't prove that `cls` is a type", and then two lines down it's saying "this function is supposed to return a qualified type `BidictBase[VT, KT]`, but I can't prove that the dynamic call to `type()` actually gave that, I just know that `inv_cls` is a `Type[_]`.

Pyright seems to have the ability to disable classes of errors on a per-file basis with a comment like `# pyright: reportGeneralTypeIssues=false`. The available categories of errors is here (https://github.com/microsoft/pyright/blob/main/docs/configuration.md#type-check-diagnostics-settings), but that's a sort of catch-all one. I'd like to be able to disable it just for this line (because it's a dynamic type creation, it makes sense why a static checker wouldn't understand it), but I can only toggle it off for the whole file.

Therefore, I recommend we switch back to pylance (because it's faster/better than Jedi IMO), but set `"python.analysis.typeCheckingMode": "off"` in devcontainer.json, and figure out why that wasn't the default for you.

Also, pyright did actually catch one type error that was real: https://github.com/leifwalsh/bidict/blob/5d8f520144e70b4da5fdcdd5f9e223f3e8e13e71/bidict/_base.py#L50 doesn't have `KT` qualified anywhere, I think it should actually be this:

```py
def _fwdm_reversed(self: 'BidictBase[KT, t.Any]`) -> t.Iterator[KT]:
    ...
```

Reply:
Shout out to #242